legate/core/legion.py

```diff
@@ -1504,17 +1654,27 @@ def get_child(self, index_partition):
         self.children[index_partition] = child
         return child
 
-    def get_root(self):
+    def get_root(self) -> Region:
         """
         Return the root Region in this tree.
         """
-        if not self.parent:
-            return self
-        return self.parent.parent.get_root()
+        if self.parent is not None:
+            return self.parent.parent.get_root()
```

Comment:
A Region has Partition-type objects as children (one for each way that the Region has been partitioned), and each Partition has Region-type objects as children (one for each sub-region in the partition). Therefore it's expected that you can can go up or down 2 steps at a time (if starting from a Region object).

Reply:
@manopapad The initializer for `Region` only sets `self.parent = parent`, but the default parameter value for `parent` is None, so it's definitely *possible* to construct a `Region` with `r.parent = None`. This is why mypy flagged the error. Given your comment, I would suggest that possibly the `parent` value to `Region` should be made non-optional. However, the `__del__` has a check that seems to indicate that None is an expected/allowed value:
```
    def __del__(self) -> None:
        if self.owned and self.parent is None:
            self.destroy(unordered=True)
```
And similar in `Region.destroy`:
```
        if self.parent is not None:
            raise RuntimeError("Only root Region objects can be destroyed")
```
So is it the case that `Region` should usually have a non-Null parent, except in specific "root" cases?

Reply:
That's correct, the root of the tree will be a `Region` with null `parent`. Every other `Region` on the same tree should have a `Partition` parent, which in turn has a non-optional `Region` parent.

Reply:
OK so there is a special case where it's *not* possible to go two steps up. I'm lacking some context to know: is the guard I added above appropriate? Or is it a problem altogether if this codepath is reached with a root region?

Reply:
The way you have it is fine. If it's easier on mypy, It would also be fine to do the following, given that `Partition` also defines `get_root()`:

```
    def get_root(self) -> Region:
        """
        Return the root Region in this tree.
        """
        if self.parent is not None:
            return self.parent.get_root()
        return self
```

Reply:
:+1: https://github.com/nv-legate/legate.core/pull/158/commits/168bedb92a1091b0469fe600cfd544b2dc5660e4